llvm/test/Transforms/InstCombine/fsqrtdiv-transform.ll

```diff
@@ -0,0 +1,406 @@
+; CHECK-SAME: double [[A:%.*]], i32 [[C:%.*]]) #[[ATTR0]] {
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[TMP0:%.*]] = call fast double @llvm.sqrt.f64(double [[A]])
+; CHECK-NEXT:    [[TMP1:%.*]] = fdiv fast double 1.000000e+00, [[A]]
+; CHECK-NEXT:    [[DIV:%.*]] = fmul fast double [[TMP1]], [[TMP0]]
+; CHECK-NEXT:    store double [[DIV]], ptr @x, align 8
+; CHECK-NEXT:    [[TOBOOL_NOT:%.*]] = icmp eq i32 [[C]], 0
+; CHECK-NEXT:    [[MUL1:%.*]] = fmul fast double [[A]], [[A]]
+; CHECK-NEXT:    [[STOREMERGE:%.*]] = select i1 [[TOBOOL_NOT]], double [[MUL1]], double [[TMP1]]
+; CHECK-NEXT:    store double [[STOREMERGE]], ptr @r1, align 8
+; CHECK-NEXT:    store double [[TMP0]], ptr @r2, align 8
+; CHECK-NEXT:    ret void
+;
+entry:
+  %0 = call fast double @llvm.sqrt.f64(double %a)
+  %div = fdiv fast double 1.000000e+00, %0
+  store double %div, ptr @x, align 8
+  %tobool.not = icmp eq i32 %c, 0
+  %mul1 = fmul fast double %a, %a
+  %mul = fmul fast double %div, %div
+  %storemerge = select i1 %tobool.not, double %mul1, double %mul
+  store double %storemerge, ptr @r1, align 8
+  %mul2 = fmul fast double %div, %a
+  store double %mul2, ptr @r2, align 8
+  ret void
+}
+
+; multiple instances of multiply ops to optimize. Optimize all.
+define void @mutiple_multiply_instances(double %a, i32 %c) #0 {
+; CHECK-LABEL: define void @mutiple_multiply_instances(
+; CHECK-SAME: double [[A:%.*]], i32 [[C:%.*]]) #[[ATTR0]] {
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[TMP0:%.*]] = tail call fast double @llvm.sqrt.f64(double [[A]])
+; CHECK-NEXT:    [[MUL:%.*]] = fdiv fast double 1.000000e+00, [[A]]
+; CHECK-NEXT:    [[DIV:%.*]] = fmul fast double [[MUL]], [[TMP0]]
+; CHECK-NEXT:    store double [[DIV]], ptr @x, align 8
+; CHECK-NEXT:    [[TOBOOL_NOT:%.*]] = icmp eq i32 [[C]], 0
+; CHECK-NEXT:    [[MUL2:%.*]] = fmul fast double [[A]], [[A]]
+; CHECK-NEXT:    [[MUL1:%.*]] = fmul fast double [[A]], [[A]]
+; CHECK-NEXT:    [[MUL_SINK:%.*]] = select i1 [[TOBOOL_NOT]], double [[MUL2]], double [[MUL]]
+; CHECK-NEXT:    [[STOREMERGE:%.*]] = select i1 [[TOBOOL_NOT]], double [[MUL]], double [[MUL1]]
+; CHECK-NEXT:    store double [[MUL_SINK]], ptr @r1, align 8
+; CHECK-NEXT:    store double [[STOREMERGE]], ptr @r3, align 8
+; CHECK-NEXT:    store double [[TMP0]], ptr @r2, align 8
+; CHECK-NEXT:    ret void
+;
+entry:
+  %0 = tail call fast double @llvm.sqrt.f64(double %a)
+  %div = fdiv fast double 1.000000e+00, %0
+  store double %div, ptr @x, align 8
+  %tobool.not = icmp eq i32 %c, 0
+  %mul2 = fmul fast double %a, %a
+  %mul3 = fmul fast double %div, %div
+  %mul = fmul fast double %div, %div
+  %mul1 = fmul fast double %a, %a
+  %mul.sink = select i1 %tobool.not, double %mul2, double %mul
+  %storemerge = select i1 %tobool.not, double %mul3, double %mul1
+  store double %mul.sink, ptr @r1, align 8
+  store double %storemerge, ptr @r3, align 8
+  %mul4 = fmul fast double %a, %div
+  store double %mul4, ptr @r2, align 8
+  ret void
+}
+
```

Comment:
Add some tests for fpmath preservation (at least 3 cases, present on sqrt, present on fdiv, and present on both).

Also add some vector cases, and negative strictfp case

Reply:
do vector cases have significance here? this pattern would kick in well before vectorization kicks in

Reply:
also, is the strictFP check required? We get all the instructions in the form of constrained intrinsics like 
%mul = tail call nnan ninf nsz arcp contract afn double @llvm.experimental.constrained.fmul.f64(double %div, double %div, metadata !"round.tonearest", metadata !"fpexcept.strict") #2

and the  'visitFdiv' function is  never visited. So, no point in having this check.

Reply:
Where vectorization happens or not does not matter, the code should handle vectors no matter what. You would need to write the strictfp test in terms of the constrained intrinsics, but it would still be good to have the negative test just in case somebody had the idea to make the matchers cover those too